tests/components/kodi/test_config_flow.py

```diff
@@ -0,0 +1,325 @@
+            result3["flow_id"], TEST_WS_PORT
+        )
+
+    assert result4["type"] == "form"
+    assert result4["errors"] == {"base": "cannot_connect"}
+
+
+async def test_form_exception_ws(hass, user_flow):
+    """Test we handle generic exception over WebSocket."""
+    with patch(
+        "homeassistant.components.kodi.config_flow.Kodi.ping", return_value=True,
+    ), patch(
+        "homeassistant.components.kodi.config_flow.get_kodi_connection",
+        return_value=MockConnection(),
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            user_flow, TEST_CREDENTIALS
+        )
+
+    with patch(
+        "homeassistant.components.kodi.config_flow.Kodi.ping", side_effect=Exception,
+    ), patch(
+        "homeassistant.components.kodi.config_flow.get_kodi_connection",
+        return_value=MockConnection(),
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"], TEST_WS_PORT
+        )
+
+    assert result2["type"] == "form"
+    assert result2["errors"] == {"base": "unknown"}
+
+
+async def test_discovery(hass, discovery_flow):
+    """Test discovery flow works."""
+    with patch(
+        "homeassistant.components.kodi.config_flow.Kodi.ping", return_value=True,
+    ), patch(
+        "homeassistant.components.kodi.config_flow.get_kodi_connection",
+        return_value=MockConnection(),
+    ), patch(
+        "homeassistant.components.kodi.async_setup", return_value=True
+    ) as mock_setup, patch(
+        "homeassistant.components.kodi.async_setup_entry", return_value=True,
+    ) as mock_setup_entry:
+        result = await hass.config_entries.flow.async_configure(
+            discovery_flow, TEST_CREDENTIALS
+        )
+
+        assert result["type"] == "form"
+        assert result["errors"] == {}
+
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"], TEST_WS_PORT
+        )
+
+    assert result2["type"] == "create_entry"
+    assert result2["title"] == "hostname"
+    assert result2["data"] == {
+        **TEST_HOST,
+        **TEST_CREDENTIALS,
+        **TEST_WS_PORT,
+        "name": "hostname",
+        "timeout": DEFAULT_TIMEOUT,
+    }
+
+    await hass.async_block_till_done()
+    assert len(mock_setup.mock_calls) == 1
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_discovery_cannot_connect_http(hass, discovery_flow):
+    """Test discovery aborts if cannot connect."""
+    with patch(
+        "homeassistant.components.kodi.config_flow.Kodi.ping",
+        side_effect=CannotConnectError,
+    ), patch(
+        "homeassistant.components.kodi.config_flow.get_kodi_connection",
+        return_value=MockConnection(),
+    ):
+        result = await hass.config_entries.flow.async_configure(
+            discovery_flow, TEST_CREDENTIALS
+        )
+
+    assert result["type"] == "abort"
+    assert result["reason"] == "cannot_connect"
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6042836)